### ISSUE TITLE
Mute / unmute check box shows wrong value when clicked + popup window overlaps panel

### ISSUE THREAD

**Issue by milos3** [nicklan/pnmixer#69]
1. The mute / unmute check box in the popup window shows the wrong value when clicking on it. It corrects itself when changing the volume slider. This happens on both gtk2 and gtk3.
![pnmixer_bug](https://cloud.githubusercontent.com/assets/5746405/7236130/4cdae172-e793-11e4-89a9-efbcde0bea2c.gif)

2. The second problem I found is that the popup window overlaps the panel instead of opening above the panel (in case of a bottom panel). This affects both gtk2 and gtk3. I do not know if this is intended or not, because I found out about pnmixer on Manjaro Linux which had an earlier version of pnmixer which didn't overlap the panel.
![pnmixer_bug_popup_before](https://cloud.githubusercontent.com/assets/5746405/7236231/ec7af4ce-e793-11e4-8c83-41cd05290328.jpg)
For gtk2 I got around this by removing the 30th line`<property name="type">popup</property>` from popup_window-gtk2.xml before compiling which did the job.
![pnmixer_bug_popup_after](https://cloud.githubusercontent.com/assets/5746405/7236341/9da1fde2-e794-11e4-8fbb-4a958192956a.jpg)

**Comment by elboulangero** [comment]
1. I played a little bit with the volume slider, and I saw plenty of funny things happening. Not exactly the same as you, but it's probably related. I think there are a few things that need to be fixed.
2. Yep, the popup window overlaps the panel, I think I always saw pnmixer like that, it never really bothered me. I don't know either if it's intended or not.

**Comment by milos3** [comment]
1. In that exact example there's nothing wrong with the volume slider.
2. I don't think it's a good design to have the popup cover the panel completely as it feels way out of place. But hey, that0s my personal opinion :).

While I'm at it, here are two more minor anoyances:

* 3.. To change the volume slider I cannot just click on a selected value and expect the slider to change to that value but instead it increases/decreases by 10% (continuously if I hold). Only tested on gtk2. Newest version 0.6-dev:
![pnmixer_drag_after](https://cloud.githubusercontent.com/assets/5746405/7244300/b7840a30-e7d9-11e4-8610-c65e8a6f7128.gif)
On Version 0.5.1 I can click anywhere and the slider will jump to that value and I can drag it anywhere:
![pnmixer_drag_before](https://cloud.githubusercontent.com/assets/5746405/7244304/ce82b060-e7d9-11e4-8aa7-42c2c753db6d.gif)

* 4.. If double clicking on a blank space on the popup menu it makes it close. Don't know if it's intended. Also if I double click on the slider the values change randomly or from 0 to 100 and vice versa. Only tested on gtk2. Newest version 0.6-dev:
![pnmixer_clicking_after](https://cloud.githubusercontent.com/assets/5746405/7244339/45dcec48-e7da-11e4-868b-ba6c9153d19c.gif)
On Version 0.5.1 it doesn't close and the slider doesn't change randomly:
![pnmixer_clicking_before](https://cloud.githubusercontent.com/assets/5746405/7244342/4e969f64-e7da-11e4-9d16-d046dd41a469.gif)

The mentioned 0.5.1 version is from Manjaro Linux. All four issues aren't major but sometimes the small thing are most important! For 3. and 4. issue I do not know if they were intended or just overlooked but in my opinion the older version feels more polished.

**Comment by elboulangero** [comment]
1. Let's try yo have more info about your soundcard. Can you type the command 'amixer' in your console please, and paste me the result here ? Also, please tell me which info you entered, in the "Preferences -> Device" part of PNMixer.

2. I also prefer the window to be outside of the panel. However, removing `<property name="type">popup</property>` as you suggest produce an error message each time the popup window is closed, so I don't know if it's the right way to do it.
Another thing to consider is that PNMixer uses GtkStatusIcon to display its icon, and it has been deprecated since Gtk 3.14. So, there will be a little bit of code re-work one of these days in the future, and maybe it will be the right time to consider this issue.

3. Indeed I noticed that, I don't know if this behavior is due to Gtk2. But with Gtk3 it works as expected.

4. Same thing, it's OK on Gtk3 version.

I also just pushed a fix related to the slider bar, so maybe you can pull the current version and try again with Gtk3 version, and tell me what you think.

**Comment by milos3** [comment]
I forgot to mention that at first I did install pnmixer on my laptop but shifted to Virtualbox because of snapshots for other programs, so I mostly tested pnmixer on Virtualbox. When you mentioned about sound cards I went ahead and tested it again on my laptop in case it was Virtualbox's fault.

1. The mute/unmute check box was indeed an issue of Virtualbox when the Card under Device was choosen as `Intel 82801AA-ICH`, but when I choose `(default)` it worked normally in Virtualbox. On my Intel laptop I don't have this issue on neither `(default)` but I did notice a strange behavour when I choose `HDA Intel`

**Comment by milos3** [comment]
I should have noted that at first I did install pnmixer on my laptop but then I shifted to Virtualbox because of shapshots for other programs so most of my testing was done in Virtualbox. When you mentioned sound cards I went ahead and tested pnmixer again on my Intel laptop to make sure it wasn't Virtualbox's fault.

1. The mute/unmute check box visual issue as indeed present only in Virtualbox but only when I have the actual sound card choosen `Intel 82801AA-ICH` under Card in Device preferences. When I choose `(default)` it works normally in Virtualbox.
On my laptop `(default)` works normally as it should, but when choosing the card `HDA Intel` in my case, when clicking on the check box, it shows the correct icon state but I still couldn't hear any sound when unmuting. I did look into this with alsamixer and found out that when muting a choosen channel it mutes **all** of the channels but when unmuting the same choosen channel it only unmutes that **one** channel. Shouldn't it only mute and unmute the choosen channel instead of muting all channels and unmuting only choosen channel?

2. There was no errors displayed in the terminal to me when I removed `<property name="type">popup</property>` from both gtk2 and gtk3. Is there a verbose paramater that I'm missing?

3. I did test gtk3 now aswell and yes it works fine with gtk3 but still the same issue on gtk2. As shown from above gif, that behaviour wasn't present in a previous version I tested 0.5.1-2 on Manjaro, which I also checked the sources on their pnmixer and it is using gtk2 indeed.

4. On gtk3 the popup window still closes on a double click on a blank space but when double clicking on the volume slider it does not change the volume "randomly". With gtk2 it's the same as before, but I should add that when double clicking on the volume slider it does not change randomly but rather it changes to the volume value that is parallel to the mouse, I suspect that's because the slider is still beeing held by the mouse when closed and beeing released when opening the popup window. The same could be said for when double clicking left/right of the slider and it going from 0 to 100 and vice versa, the slider is beeing held when closed because of the double click.

5. One more thing I noticed when installing pulseaudio is that the slider gets very lagging when draging. Only affects gtk2, gtk3 is fine.
![pnmixer_draging_laggy](https://cloud.githubusercontent.com/assets/5746405/7339866/803d78d8-ec7d-11e4-837b-ce8f303d5aa3.gif)

All above I tested with my laptop aswell and only in the 1. issue there was a slight difference which I mentioned. Tested on the newest git version.

**Comment by elboulangero** [comment]
1. I checked that with alsamixer as well, and everything is fine on my laptop. Only the selected channel gets muted/unmuted. Looking at the code, the 'mute' and 'unmute' code is exactly the same, just replace a one by a zero. It's hard to figure out what's going wrong here.

2. Nope. The error message only appears in GTK3 version. Just show/hide the popup window, and witness the line:
`(pnmixer:16913): Gdk-CRITICAL **: gdk_device_get_window_at_position_double: assertion 'GDK_IS_DEVICE (device)' failed`
Now, I did a little bit of investigation and found this thread: https://github.com/nicklan/pnmixer/issues/37 which lead to this pull request https://github.com/nicklan/pnmixer/pull/42.
What I understand from it: in the 0.5 version, PNMixer didn't have the 'popup' property. It looked good on some window manager, but was broken on others. Apparently it has been quite a pain to fix it for everyone. Now, with the 'popup'property, it works fine, but alas, it appears on top of the panel. And honestly, I have no idea how it can be changed.

3. I got the version of pnmixer 0.5.1-3, from Archlinux, compiled it, and it works as in the git version. That is, I click on the slider bar, and it increases/decreases by 10%. Don't know, maybe the Manjaro version of PNMixer differs from the one available in Arch. Which version of Manjaro do you use btw ?

4. I agree that double-clicking on the slider bar in GTK2 is bogus. However, by comparing quickly between GTK2 and GTK3 version, there's nothing that strikes me, it's very similar. And there's no problem in GTK3, so it may be a GTK2 bug. And for the difference with 0.5 version, it comes from the fix mentioned in the point 2. Maybe @hasufell or @bbidulock can help you with that.

5. Hmmmm, well I personally don't use pulseaudio, and as far as I know PNMixer is thought for people who use ALSA. I would say that using PNMixer and Pulse is 'undefined behavior', or 'use at your own risk', or call it whatever you want ;)

Now I run out of time, I'll be back again next week-end

**Comment by elboulangero** [comment]
Hi Milos,
can you please open several tickets, as it becomes difficult so read this one, please ? Sorry I should have asked that from the beginning. I can do it if you want, but I think it's better if you do, so that history remembers who reported these issues.

You can just copy/paste what's interesting, no need to rewrite everything.
My proposal:
1. One ticket for the mute/unmute issue. I'm gonna create a branch with more debug messages, so that you can test and send me the output.
2. One for the popup that appears above the panel. You can tag that at both gtk2/3.
3. One for the issue 3, tagged as gtk2 only.
4. The issue 4 you can split in two
 - one for the weird behavior of the bogus behavior of the scrollbar, tagged as gtk2 only
 - one for the fact that double-click on the popup window closes it, gtk2/3
5. At last, you can create a ticket for the pulseaudio issue

For the issues 4 I've got a fix, but I wait for you to open the tickets. If you don't want to do that, just say it I'll do it, no problem.

THanks !

**Comment by milos3** [comment]
Sure thing! I'll start opening the tickets in a few hours when I come home. Yea it is hard to follow them all in a single issue, sorry about that.

**Comment by milos3** [comment]
Alright, I opened all of the tickets now.

In the 1. issue the check box bug is only affected under virtualbox when I choose `ICH AC97` Audio Controller under Audio in virtualbox Settings and `Intel 82801AA-ICH` Device Card in pnmixer settings. When Intel `HD Audio` is choosen in virtualbox Settings instead then I don't have that bug. I did not open a new ticket regarding this as it may be virtualbox's fault. But I did open regarding muting/unmuting a single channel, which turned out to be pulseaudio bug, which makes that ticket a possible duplicate to https://github.com/nicklan/pnmixer/issues/17.
